### PR TITLE
Added new params to file copy

### PR DESCRIPTION
**Pull request by khazamov**

### CODE REVIEW

File: pyuploadcare/api_resources.py [hunk 1]
```diff
@@ -231,12 +231,30 @@ def store(self):
...
         """Creates File copy
 
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
+      - If ``target`` is ``None``, copy file to Uploadcare storage otherwise
+        copy to target associated with project;
+      -  Add ``effects`` to ``self.default_effects`` if any;
+      -  To grant a file public access on the target storage set
+        ``make_public`` option to be ``true``,
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: true is default behavior

File: pyuploadcare/api_resources.py [hunk 2]
```diff
@@ -231,12 +231,30 @@ def store(self):
...
+        relevant when ``target`` has a value set.
+      -  On the contrary, ``store`` option is actual only when ``target`` is ``None``.
+        Stores file permanently in your Uploadcare storage.
+        To avoid confusion, please make sure your account auto store settings
+        do not overlap with this option.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: this is redundant here

File: pyuploadcare/api_resources.py [hunk 3]
```diff
@@ -231,12 +231,30 @@ def store(self):
...
+      - Specify  ``target`` option to set the file name
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: ``pattern``

File: pyuploadcare/api_resources.py [hunk 4]
```diff
@@ -231,12 +231,30 @@ def store(self):
...
+      Following example copies a file to ``samplefs`` S3 storage with an owner only access,
+      with file name consisting of uuid,original filename and file extension placed
+      in the root of the storage.
+      - >>>file.copy(target='samplefs',
+                    make_public='false',
+                    pattern='${uuid}${filename}${ext}');
+      In this example, a file is copied to 'samplefs' storage with public access and placed
+      under the directory named after file uuid. A new file has the same file name as its original.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: it's not true, I think it'll be uuidfilename.ext

File: pyuploadcare/api_resources.py [hunk 5]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
+      - If ``target`` is ``None``, copy file to Uploadcare storage otherwise
+        copy to target associated with project;
+      -  Add ``effects`` to ``self.default_effects`` if any;
+      -  To forbid a public from accessing your files on the target storage set
+        ``make_public`` option to be ``false``,
+        relevant when ``target`` has a value set.
+      - Specify  ``pattern`` option to set the file name
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: file name -> s3 object key name

File: pyuploadcare/api_resources.py [hunk 6]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
+
+      - Following example copies a file to ``samplefs`` S3 storage with an owner only access,
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: "owner only"

File: pyuploadcare/api_resources.py [hunk 7]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
+      with file name consisting of uuid,original filename and file extension placed
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: same as above

add space after comma

File: pyuploadcare/api_resources.py [hunk 8]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
+      in the root of the storage:
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: of the bucket

File: pyuploadcare/api_resources.py [hunk 9]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
+            - >>>file.copy(target='samplefs',
+                            make_public='false',
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: `make_public=False`

File: pyuploadcare/api_resources.py [hunk 10]
```diff
@@ -231,12 +231,28 @@ def store(self):
...
+                            pattern='${uuid}${filename}${ext}');
+
+      - In this example, a file is copied to 'samplefs' storage with public access and placed
+      under the directory named after file uuid. A new file has the same file name as its original.
+                >>>file.copy(None,'samplefs','true');
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: 'true' is redundant

File: pyuploadcare/api_resources.py (hunk 5 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: как мы называем это в случае копирования внутри uploadcare? file name или object key? какое из них более общее?

File: pyuploadcare/api_resources.py (hunk 10 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: pyuploadcare/api_resources.py (hunk 9 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: pyuploadcare/api_resources.py (hunk 5 above)
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: currently we're working with S3, so: http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html

also please use English in public repo :)

Comment (khazamov) [inline pyuploadcare/api_resources.py]: that's right, but the term "file" is used throughout the docs

Comment (khazamov) [inline pyuploadcare/api_resources.py]: yep

**Comment by dmitry-mukhin** [comment]
[file.copy в pyuploadcare , phplib](https://trello.com/c/blfzlKNg/164-file-copy-pyuploadcare-phplib)

File: pyuploadcare/api_resources.py [hunk 11]
```diff
@@ -231,12 +231,28 @@ def store(self):
         """
         self._info_cache = rest_request('PUT', self._api_storage_uri)
 
-    def copy(self, effects=None, target=None):
+    def copy(self, effects=None, target=None, make_public=None, store=None, pattern=None):
         """Creates File copy
```
Comment (sharov) [inline pyuploadcare/api_resources.py]: I suggest to reformat function definition like this

def copy(self, effects=None, store=None, target=None, make_public=None, pattern=None):

And to change the order of description of params in docstring, in accordance with function definition. Of course, description of ‘store’ is needed.

Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: this will break backwards compatibility (as I have learned the hard way). moving `store` to the fourth place from fifth is reasonable though. `make_public` can be the last as I think that `pattern` will be used more.

Comment (andrewshkovskii) [inline pyuploadcare/api_resources.py]: Since number of parameters are grew - we can use kwargs, so no need for positional paramaters. 
Just write the good docs and everything will be fine.

Comment (khazamov) [inline pyuploadcare/api_resources.py]: using kwargs sounds good.

File: pyuploadcare/api_resources.py [hunk 12]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
-        """
-        effects = effects or ''
+        Keyword Args:
+          - target:  
+                If ``target`` is ``None``, copy file to Uploadcare storage otherwise
+                copy to target associated with project;
+          - effects: 
+                Add ``effects`` to ``self.default_effects`` if any;
+          - make_public: 
+                To forbid a public from accessing your files on the target storage set
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: to forbid public access

File: pyuploadcare/api_resources.py [hunk 13]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
+                ``make_public`` option to be False, relevant when ``target`` has a value set.
+          - store: 
+                If ``store`` option is set to False the copy of your file will be deleted
+                in 24 hour period after the upload.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: only works with Uploadcare storage

File: pyuploadcare/api_resources.py [hunk 14]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
+          - pattern: 
+                Specify  ``pattern`` option to set S3 object key name.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: works with custom storage and overrides pattern set in storage settings.

File: pyuploadcare/api_resources.py [hunk 15]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
+        and file name consisting of uuid concatenated to original filename  placed
+        in the root of the bucket:
+             >>>file.copy(target='samplefs',
+                                make_public=False,
+                                pattern='${uuid}${filename}${ext}');
+
+        In this example, a file is copied to ``samplefs`` storage with public access and placed
+        under the directory named after file uuid. A new file has the same file name as its original.
+                >>>file.copy(None,'samplefs');
+
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: you can borrow descriptions directly from https://uploadcare.com/documentation/rest/#file

File: pyuploadcare/api_resources.py [hunk 16]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
+        """
+        effects = kwargs.get('effects','')
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: add space after comma

File: pyuploadcare/api_resources.py [hunk 17]
```diff
@@ -231,25 +231,45 @@ def store(self):
...
             effects = fmt.format(head=self.default_effects,
                                  tail=effects)
-        data = {
-            'source': self.cdn_path(effects)
-        }
-        if target is not None:
-            data['target'] = target
 
-        return rest_request('POST', 'files/', data=data)
+        kwargs.update({'source':self.cdn_path(effects)})
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: add space after colon

File: pyuploadcare/api_resources.py (hunk 12 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 13 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 14 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 15 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 16 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 17 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py [hunk 18]
```diff
@@ -231,25 +231,52 @@ def store(self):
...
-    def copy(self, effects=None, target=None):
+    def copy(self, **kwargs):
         """Creates File copy
-
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
-        """
-        effects = effects or ''
+        Keyword Args:
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: we need empty line here, afair

**Comment by khazamov** [comment]
All my local tests are passing.

**Comment by homm** [comment]
> Since number of parameters are grew - we can use kwargs

I don't agree. `kwargs` makes spelling errors possible and also breaks compatibility.

**Comment by homm** [comment]
Positional arguments should be returned at that stage.

For long terms we should create two separate copy methods. Unfortunately, we have confusing API on REST side, but we don't have to copy them as is. In fact, we have two copy operations: copy to UC storage and copy to some target. For the local copy we need `effects` and `store` arguments, for the remote copy we need `effects `, `make_public`, `pattern`.

From that point of view, it is better to group `target`, `make_public` and `pattern` arguments together. And make `store` the last.

**Comment by homm** [comment]
Created the issue #119

File: pyuploadcare/api_resources.py [hunk 19]
```diff
@@ -231,53 +231,77 @@ def store(self):
...
-                ``make_public`` option to be False, relevant when ``target`` has a value set.
-                Default value is True.
-          - store:
-                If ``store`` option is set to False the copy of your file will be deleted
-                in 24 hour period after the upload. Relevant to Uploadcare storage only.
-          - pattern: 
-                Specify  ``pattern`` option to set S3 object key name. Takes precedence over
-                pattern set in the settings. If neither is specified defaults to ${uuid}/${auto_filename}.
-                Relevant to custom storage only.
+    def lcopy(self, effects=None, store=None):
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: не нравятся названия.
в питоне rxxxx lxxxx обычно значит right, left.

варианты:
`local_copy, remote_copy`
`copy, copy_to`

надо ещё подумать

File: pyuploadcare/api_resources.py [hunk 20]
```diff
@@ -231,53 +231,77 @@ def store(self):
...
+            effects = fmt.format(head=self.default_effects,
+                                 tail=effects)
+        data = {
+            'source': self.cdn_path(effects)
+        }
+        if store is not None:
+            data['store'] = store
+        return rest_request('POST', 'files/', data=data)
+
+    def rcopy(self, effects=None, target=None, make_public=None, pattern=None):
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: тут таргет явно на первое место просится и он должен быть обязательны параметром

File: pyuploadcare/api_resources.py (hunk 19 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: addressed

File: pyuploadcare/api_resources.py (hunk 20 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/dj/conf.py [hunk 21]
```diff
@@ -24,7 +24,7 @@
     conf.cdn_base = settings.UPLOADCARE['cdn_base']
 
 
-widget_version = settings.UPLOADCARE.get('widget_version', '2.10.2')
+widget_version = settings.UPLOADCARE.get('widget_version', '2.10.0')
```
Comment (sharov) [inline pyuploadcare/dj/conf.py]: Are you sure?

Comment (khazamov) [inline pyuploadcare/dj/conf.py]: We agreed that splitting copy methods to local and remote  would be a major change, and this branch would go into master along with the other changes introduced in the future. So the version change is symbolic. Not sure how to go about pull request though.

Comment (khazamov) [inline pyuploadcare/dj/conf.py]: widget version should be reverted, the above still holds true

File: pyuploadcare/api_resources.py [hunk 22]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+            Relevant to custom storage only.
+
+        For more information on each of the options above please refer to
+        our docs https://uploadcare.com/documentation/rest/#file.
+
+        Following example copies a file to S3 bucket named ``samplefs`` with  owner only access,
+        and file name consisting of UUID concatenated to original filename. The file is placed
+        in the root of the storage:
+             >>>file.remote_copy(target='samplefs',
+                                make_public=False,
```
Comment (homm) [inline pyuploadcare/api_resources.py]: add one indention space

File: pyuploadcare/api_resources.py [hunk 23]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+        if store is not None:
+            data['store'] = store
+        return rest_request('POST', 'files/', data=data)
+
+    def remote_copy(self, target, effects=None, make_public=None, pattern=None):
+        """Creates File Copy On Remote Storage
+
+        - effects:
+            Adds CDN image effects to ``self.default_effects`` if any;
+        - target:
```
Comment (homm) [inline pyuploadcare/api_resources.py]: Target is the first in arguments order.

File: pyuploadcare/api_resources.py [hunk 24]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+            If ``target`` is ``None``, copy file to Uploadcare storage otherwise
+            copy to target associated with project;
+        - make_public:
+            To forbid a public access for your files on the target storage set
+            ``make_public`` option to be False, relevant when ``target`` has a value set.
```
Comment (homm) [inline pyuploadcare/api_resources.py]: > relevant when ``target`` has a value set

Target is mandatory, it always has a value.

File: pyuploadcare/api_resources.py [hunk 25]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+            Default value is True.
+        - pattern:
+            Specify  ``pattern`` option to set S3 object key name. Takes precedence over
+            pattern set in the settings. If neither is specified defaults to ${uuid}/${auto_filename}.
+            Relevant to custom storage only.
+
+        For more information on each of the options above please refer to
+        our docs https://uploadcare.com/documentation/rest/#file.
+
+        Following example copies a file to S3 bucket named ``samplefs`` with  owner only access,
```
Comment (homm) [inline pyuploadcare/api_resources.py]: > to S3 bucket named ``samplefs``

Target name and bucket named can be differ.

File: pyuploadcare/api_resources.py [hunk 26]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+            data['store'] = store
+        return rest_request('POST', 'files/', data=data)
+
+    def remote_copy(self, target, effects=None, make_public=None, pattern=None):
+        """Creates File Copy On Remote Storage
+
+        - effects:
+            Adds CDN image effects to ``self.default_effects`` if any;
+        - target:
+            If ``target`` is ``None``, copy file to Uploadcare storage otherwise
```
Comment (homm) [inline pyuploadcare/api_resources.py]: Target is mandatory. Please, update this.

File: pyuploadcare/api_resources.py [hunk 27]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+
+        For more information on each of the options above please refer to
+        our docs https://uploadcare.com/documentation/rest/#file.
+
+        Following example copies a file to S3 bucket named ``samplefs`` with  owner only access,
+        and file name consisting of UUID concatenated to original filename. The file is placed
+        in the root of the storage:
+             >>>file.remote_copy(target='samplefs',
+                                make_public=False,
+                                pattern='${uuid}${filename}${ext}');
```
Comment (homm) [inline pyuploadcare/api_resources.py]: Maybe better add slash? `'${uuid}/${filename}${ext}'`

File: pyuploadcare/api_resources.py [hunk 28]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+
+        In this example, a file is copied to ``samplefs`` storage. Options ``make_public`` and ``pattern``
+        are not set, so default settings  will kick in: the file has public access rights and it is placed
+        in the root of the directory named after file UUID. A new file has the same file name as its original.
+                >>>file.remote_copy(None,'samplefs');
```
Comment (homm) [inline pyuploadcare/api_resources.py]: What is it?

File: pyuploadcare/api_resources.py [hunk 29]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+        effects = effects or ''
+        if self.default_effects is not None:
+            fmt = '{head}-/{tail}' if effects else '{head}'
+            effects = fmt.format(head=self.default_effects,
+                                 tail=effects)
+        data = {
+            'source': self.cdn_path(effects)
+        }
+
         if target is not None:
```
Comment (homm) [inline pyuploadcare/api_resources.py]: target is mandatory

File: pyuploadcare/api_resources.py (hunk 27 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: wanted to demonstrate the filename can be set to a different value from the default one

File: pyuploadcare/api_resources.py [hunk 30]
```diff
@@ -231,24 +231,76 @@ def store(self):
         """
         self._info_cache = rest_request('PUT', self._api_storage_uri)
 
-    def copy(self, effects=None, target=None):
-        """Creates File copy
+    def local_copy(self, effects=None, store=None):
```
Comment (khazamov) [inline pyuploadcare/api_resources.py]: yes, it should be

File: pyuploadcare/api_resources.py (hunk 27 above)
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: default one is ${default}, we just want example to make some sense.
`e8ebfe20-8c11-4a94-9b40-52ecad7d8d1a/billmurray.jpg` makes more sense than`somefolder/e8ebfe20-8c11-4a94-9b40-52ecad7d8d1abillmurray.jpg`

File: pyuploadcare/api_resources.py (hunk 30 above)
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: what about `create_local_copy`/`create_remote_copy`

File: pyuploadcare/__init__.py [hunk 31]
```diff
@@ -15,7 +15,7 @@
 
 from __future__ import unicode_literals
 
-__version__ = '2.1.0'
+__version__ = '3.0.0'
```
Comment (khazamov) [inline pyuploadcare/__init__.py]: Bumping to 3.0.0  with deprecation warning for 4.x.x. based on the changelog.  What is your suggestion?

File: pyuploadcare/api_resources.py (hunk 30 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: that's more in line with the code in WC, while @homm's idea might be better aligned with lib's history. Let's go with copy_to_remote and copy_to_local.

File: pyuploadcare/api_resources.py [hunk 32]
```diff
@@ -231,24 +231,76 @@ def store(self):
...
+        """Creates a Local File Copy on Uploadcare Storage
 
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
+        - effects:
+            Adds CDN image effects to ``self.default_effects`` if any;
+        - store:
+            If ``store`` option is set to False the copy of your file will be deleted
+            in 24 hour period after the upload.
```
Comment (khazamov) [inline pyuploadcare/api_resources.py]: Changed wording. Actual description is  given in definition of cdn_url function.

File: pyuploadcare/api_resources.py (hunk 23 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 26 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 24 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 25 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 22 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 27 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/__init__.py (hunk 31 above)
Comment (dmitry-mukhin) [inline pyuploadcare/__init__.py]: just revert the change. the scope of this PR is adding two new methods to File resource, that's it.

File: pyuploadcare/api_resources.py [hunk 33]
```diff
@@ -231,24 +231,81 @@ def store(self):
         """
         self._info_cache = rest_request('PUT', self._api_storage_uri)
 
-    def copy(self, effects=None, target=None):
-        """Creates File copy
+    def copy(self, effects=None, target=None, make_public=None, store=None, pattern=None):
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: do not change the signature. for backward compatibility we'll leave this as is.

File: pyuploadcare/api_resources.py [hunk 34]
```diff
@@ -231,24 +231,81 @@ def store(self):
...
+        """Creates a File Copy on Uploadcare or Custom Storage. Will be deprecated in version 4.0.0"""
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: raise deprecation warning here

File: setup.py [hunk 35]
```diff
@@ -23,7 +23,7 @@
 
 setup(
     name='pyuploadcare',
-    version='2.1.0',
+    version='3.0.0',
```
Comment (dmitry-mukhin) [inline setup.py]: don't change the version here/

File: HISTORY.rst [hunk 36]
```diff
@@ -3,13 +3,16 @@
 History
 -------
 
-x.x.x
+3.0.0
```
Comment (homm) [inline HISTORY.rst]: Also should not be changed there

File: pyuploadcare/api_resources.py [hunk 37]
```diff
@@ -231,24 +231,81 @@ def store(self):
...
 
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
+        if target is not None:
```
Comment (homm) [inline pyuploadcare/api_resources.py]: Maybe invert the condition?

File: pyuploadcare/api_resources.py [hunk 38]
```diff
@@ -231,24 +231,81 @@ def store(self):
...
+        """Creates File Copy On Remote Storage
+
+        - target:
+            Target name associated with the project;
+        - effects:
+            Adds CDN image effects to ``self.default_effects`` if any;
+        - make_public:
+            To forbid public from accessing your files on the storage set
+            ``make_public`` option to be False.
+            Default value is True.
```
Comment (homm) [inline pyuploadcare/api_resources.py]: But it is None )

File: pyuploadcare/api_resources.py [hunk 39]
```diff
@@ -231,24 +231,81 @@ def store(self):
...
+        - pattern:
+            Specify  ``pattern`` option to set S3 object key name. Takes precedence over
```
Comment (homm) [inline pyuploadcare/api_resources.py]: Extra space after Specify

File: pyuploadcare/api_resources.py [hunk 40]
```diff
@@ -231,24 +231,81 @@ def store(self):
...
+            pattern set in the settings. If neither is specified defaults to
```
Comment (homm) [inline pyuploadcare/api_resources.py]: in the project settings?

**Comment by homm** [comment]
> Let's discuss the names of new methods one more time, copy_to_x is cringy.

create_local_copy/create_remote_copy looks good.

File: pyuploadcare/api_resources.py (hunk 38 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: pyuploadcare/api_resources.py (hunk 37 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: api_resources is full of inverted conditions. Maybe include in refactoring list to-do?

File: pyuploadcare/api_resources.py (hunk 39 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: pyuploadcare/api_resources.py (hunk 40 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: fixed

File: HISTORY.rst (hunk 36 above)
Comment (khazamov) [inline HISTORY.rst]: fixed

File: setup.py (hunk 35 above)
Comment (khazamov) [inline setup.py]: fixed

File: pyuploadcare/api_resources.py (hunk 34 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py (hunk 33 above)
Comment (khazamov) [inline pyuploadcare/api_resources.py]: done

File: pyuploadcare/api_resources.py [hunk 41]
```diff
@@ -232,23 +232,85 @@ def store(self):
...
+        with original filename billmurray.jpg in
+        in the directory named ``e8ebfe20-8c11-4a94-9b40-52ecad7d8d1a``.
+
+
+        """
+        effects = effects or ''
+        if self.default_effects is not None:
+            fmt = '{head}-/{tail}' if effects else '{head}'
+            effects = fmt.format(head=self.default_effects,
+                                 tail=effects)
```
Comment (sharov) [inline pyuploadcare/api_resources.py]: It seems there is copy of this code here https://github.com/uploadcare/pyuploadcare/pull/116/files#diff-2e1fdb139bd11faec8328901cf61df9aL242
I suggest making a method from this code

**Comment by andrewshkovskii** [comment]
It seems to me that `copy_to_target` can be more obvious than `create_remote_copy`.
But `create_local/remote_copy` is ok too.
But, anyway - I refactored some docstrings and added method about @sharov talked about.

File: pyuploadcare/api_resources.py [hunk 42]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
-        If ``target`` is ``None``, copy file to Uploadcare storage otherwise
-        copy to target associated with project.
-        Add ``effects`` to ``self.default_effects`` if any.
         """
-        effects = effects or ''
-        if self.default_effects is not None:
-            fmt = '{head}-/{tail}' if effects else '{head}'
-            effects = fmt.format(head=self.default_effects,
-                                 tail=effects)
+        warning = """ File.copy method will be deprecated in version 4.0.0.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: is deprecated and will be removed in 4.0.0

File: pyuploadcare/api_resources.py [hunk 43]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+    def create_local_copy(self, effects=None, store=None):
+        """Creates a Local File Copy on Uploadcare Storage.
+
+        Args:
+            - effects:
+                Adds CDN image effects. If ``self.default_effects`` property
+                is set effects will be combined with default effects.
+            - store:
+                If ``store`` option is set to False the copy of your file will
+                be deleted in 24 hour period after the upload.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: works only if autostore is enabled in the project

File: pyuploadcare/api_resources.py [hunk 44]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+
+        """
+        effects = self._build_effects(effects)
+        store = store or ''
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: `store or None` or ...

File: pyuploadcare/api_resources.py [hunk 45]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
         data = {
             'source': self.cdn_path(effects)
         }
-        if target is not None:
-            data['target'] = target
+        if store is not None:
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: ... or `if store`

File: pyuploadcare/api_resources.py [hunk 46]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+            data['store'] = store
+        return rest_request('POST', 'files/', data=data)
+
+    def create_remote_copy(self, target, effects=None, make_public=None,
+                           pattern=None):
+        """Creates file copy on remote storage.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: on -> in

File: pyuploadcare/api_resources.py [hunk 47]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
 
+        Args:
+            - target:
+                Name of a custom storage connected to your project.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: to the project

File: pyuploadcare/api_resources.py [hunk 48]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+                Adds CDN image effects to ``self.default_effects`` if any.
+            - make_public:
+                To forbid public from accessing your files on the storage set
+                ``make_public`` option to be False.
+                Default value is None. Files have public access by default.
+            - pattern:
+                Specify ``pattern`` option to set S3 object key name.
+                Takes precedence over pattern set in project settings.
+                If neither is specified defaults to
+                ${uuid}/${filename} ${effects} ${ext}.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: whitespaces ?

File: pyuploadcare/api_resources.py [hunk 49]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+
+        For more information on each of the options above please refer to
+        our docs https://uploadcare.com/documentation/rest/#file.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: our -> REST API

File: pyuploadcare/api_resources.py [hunk 50]
```diff
@@ -232,23 +238,93 @@ def store(self):
...
+
+        Following example copies a file to custom storage named ``samplefs``:
+
+             >>> file = File('e8ebfe20-8c11-4a94-9b40-52ecad7d8d1a')
+             >>> file.remote_copy(target='samplefs',
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: `create_remote_copy` :)

File: pyuploadcare/api_resources.py (hunk 48 above)
Comment (andrewshkovskii) [inline pyuploadcare/api_resources.py]: Same as in the docs

File: pyuploadcare/api_resources.py (hunk 44 above)
Comment (andrewshkovskii) [inline pyuploadcare/api_resources.py]: Not sure if i get your right, but do you mean: `store = store or None`?

**Comment by dmitry-mukhin** [comment]
@andrewshkovskii github hides my comments, this is not finished, right?

**Comment by andrewshkovskii** [comment]
@dmitry-mukhin No, i finished.

File: pyuploadcare/api_resources.py [hunk 51]
```diff
@@ -232,23 +238,96 @@ def store(self):
         self._info_cache = rest_request('PUT', self._api_storage_uri)
 
     def copy(self, effects=None, target=None):
-        """Creates File copy
+        """Creates a File Copy on Uploadcare or Custom Storage.
+        This method Will be deprecated in version 4.0.0.
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: same as in docstring

File: pyuploadcare/api_resources.py [hunk 52]
```diff
@@ -298,17 +301,17 @@ def create_remote_copy(self, target, effects=None, make_public=None,
                 Specify ``pattern`` option to set S3 object key name.
                 Takes precedence over pattern set in project settings.
                 If neither is specified defaults to
-                ${uuid}/${filename} ${effects} ${ext}.
+                ${uuid}/${filename} ${effects} ${ext} .
```
Comment (dmitry-mukhin) [inline pyuploadcare/api_resources.py]: ? remove whitespaces please

**Comment by andrewshkovskii** [comment]
Now?: )
